pkg/causality/internal/node.go

```diff
@@ -64,204 +78,180 @@ type Node struct {
+// We are using sync.Pool to lessen the burden of GC.
+func (n *Node) Free() {
+	if n.id == invalidNodeID {
+		panic("double free")
+	}
+
+	n.id = invalidNodeID
+	n.OnResolved = nil
+	n.RandWorkerID = nil
+
+	nodePool.Put(n)
 }
 
-// OnNoConflict guarantees that fn is called
-// when the node has either no unfinished dependencies
-// or all unfinished dependencies that the node has are
-// assigned to the same worker.
-func (n *Node) OnNoConflict(fn func(id workerID)) {
+// assignTo assigns a node to a worker. Returns `true` on success.
+func (n *Node) assignTo(workerID int64) bool {
 	n.mu.Lock()
 	defer n.mu.Unlock()
-
-	if n.onResolved != nil || n.resolved.Load() {
-		panic("OnNoConflict is already called")
+	if n.assignedTo != unassigned {
+		// Already resolved by some other guys.
+		return false
 	}
 
-	workerNum, ok := n.tryResolve()
-	if ok {
-		n.resolved.Store(true)
-		fn(workerNum)
-		return
+	n.assignedTo = workerID
+	if n.OnResolved != nil {
+		n.OnResolved(workerID)
+		n.OnResolved = nil
 	}
 
-	n.onResolved = fn
-}
-
-// notifyMaybeResolved must be called with n.mu taken.
-// It should be called if n.conflictCounts is updated.
-func (n *Node) notifyMaybeResolved() {
-	workerNum, ok := n.tryResolve()
-	if !ok {
-		return
+	if n.dependers != nil {
+		n.dependers.Ascend(func(node *Node) bool {
+			resolvedDependees := stdAtomic.AddInt32(&node.resolvedDependees, 1)
+			stdAtomic.StoreInt64(&node.resolvedList[resolvedDependees-1], n.assignedTo)
+			node.maybeResolve(resolvedDependees, 0)
+			return true
+		})
 	}
+	return true
+}
 
-	if n.onResolved != nil {
-		if !n.resolved.Swap(true) {
-			n.onResolved(workerNum)
+func (n *Node) maybeResolve(resolvedDependees, removedDependees int32) {
+	if workerNum, ok := n.tryResolve(resolvedDependees, removedDependees); ok {
+		if workerNum < 0 {
+			panic("Node.tryResolve must return a valid worker ID")
 		}
+		n.assignTo(workerNum)
 	}
+	return
 }
 
 // tryResolve must be called with n.mu locked.
 // Returns (_, false) if there is a conflict,
-// returns (-1, true) if there is no conflict,
+// returns (rand, true) if there is no conflict,
 // returns (N, true) if only worker N can be used.
-func (n *Node) tryResolve() (int64, bool) {
-	conflictNumber := len(n.conflictCounts)
-	if conflictNumber == 0 {
-		// No conflict at all
-		return -1, true
+func (n *Node) tryResolve(resolvedDependees, removedDependees int32) (int64, bool) {
+	if n.totalDependees == 0 {
+		// No conflicts, can select any workers.
+		return n.RandWorkerID(), true
 	}
-	if conflictNumber == 1 {
-		_, ok := n.conflictCounts[unassigned]
-		if ok {
-			// All conflicts are unassigned. So
-			// no resolution is available.
-			return 0, false
+
+	if resolvedDependees == n.totalDependees {
+		// NOTE: We don't pick the last unremoved worker because lots of tasks can be
+		// attached to that worker after a time.
+		if n.totalDependees == 1 {
+			return stdAtomic.LoadInt64(&n.resolvedList[0]), true
```

Comment:
what's special about `n.resolvedList[0]`?

Reply:
Just because `totalDependees == 1`. So it is the only dependency.